ktor-server/ktor-server-plugins/ktor-server-rate-limit/jvmAndNix/src/io/ktor/server/plugins/ratelimit/RateLimiter.kt

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2014-2022 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.server.plugins.ratelimit
+
+import io.ktor.util.date.*
+import kotlinx.atomicfu.*
+import kotlin.time.*
+
+/**
+ * An interface for rate limiters.
+ */
+public interface RateLimiter {
+
+    /**
+     * Tries to consume [tokens] amound of tokens and returns state of the rate limiter.
+     */
+    public suspend fun tryConsume(tokens: Int = 1): State
+
+    /**
+     * State of the rate limiter.
+     */
+    public sealed class State {
+        /**
+         * Rate limiter has enough tokens.
+         */
+        public class Available(
+            public val remainingTokens: Int,
+            public val limit: Int,
+            public val refillAtTimeMillis: Long
```

Comment:
may be use `Duration` instead of `Long` with `Millis` suffix everywhere?

Reply:
This one represents not duration, but a moment in time. There is no stable multiplatform API for that. We have our own `GMTDate`, but it's not best to use it in this case. 
But I changed `Exhausted.toWait` property to `Duration`

Reply:
oh, correct, missed that. works for me